### ISSUE TITLE
Add a command/flag to continually display-message

### ISSUE THREAD

**Issue by mohkale** [tmux/tmux#2598]
### Issue description

This might be out of scope for tmux as a project but I'm gonna ask for it anyways.

At the moment tmux has a `display-message` command that can be used to print the result of some format string to stdout. I'd like to ask for an extension to this command either as a new-flag for it or as a new command altogether which can re-calculate and print that message at a set interval.

The use case for this is because in my status-line config I've got a little icon (next to the date) which changes depending on the active panes command.

![demo](https://media.githubusercontent.com/media/mohkale/dotfiles/b7a64e624777545fe28f33466f56c236398db1a9/.github/main.png)

The way I've implemented this at the moment is to have a [script](https://github.com/mohkale/dotfiles/blob/b7a64e624777545fe28f33466f56c236398db1a9/prog/tmux/cmds/tmux-icons) periodically run `tmux display -t "$session" -p '#{pane_current_command}'` which I then pipe elsewhere to calculate an appropriate icon to display on the status-line.

My current approach works well enough (although I'm not a fan of spawning so many tmux commands continuously) however it isn't very convenient or maintainable. Simply looking at my message-log the vast majority of logs seem to just be me querying this value.

![Screenshot_20210305_045638](https://user-images.githubusercontent.com/23294780/110069230-45c03b80-7d6f-11eb-8b5a-d468fdd51440.png)

I'd like to replace my script with a single call to `tmux display-message`, maybe of the form `tmux display-message -r 5 -t "$session" -p '#{pane_current_command}` which every 5 seconds (`-r` for repeat cause there's already a `-I` flag which I don't want to confuse with) prints out the value of the format string passed to it. This would mean the command should never exit, which personally I'm fine with but there may be complications in tmux itself preventing that.

Thoughts?

### Required information

Please provide the following information:

* tmux version (`tmux -V`): tmux next-3.3
* Platform (`uname -sp`): Linux unknown
* $TERM inside and outside of tmux (`echo $TERM`): st-256color (outside) screen-24bit (inside)
* Logs from tmux (`tmux kill-server; tmux -vv new`). Not really necessary here.

**Comment by nicm** [comment]
Hmm, possibly. Apart from filling up the message log, I'm not sure what harm this is doing as you have it? Running tmux every five seconds is not that much really.

**Comment by nicm** [comment]
I was going to say you could use #() with a check to only run it if it has changed but that won't work because you can't put #{} inside #() inside #{}, maybe we should fix that.

You could just do #(my script.sh #{pane_current_command}) and tmux will not run it too often if it hasn't changed, but you will have to live with a short delay before it sorts itself out in new panes.

**Comment by mohkale** [comment]
@nicm

All fair points. I've kind of gone out of my way to streamline everything and minimise delays as much as possible. I have a config file with icons which is read only once into an awk process which continually reads program names and dumps out icons. That's one of the reasons I'm not using the status line directly and am instead piping the program name through to my icon selector. If I did `choose-icon #{pane_command}` then I'd have to re-read the config every update/change which I felt would be too inefficient.

The icon application was really only my use case here, I'm sure there're other more meaningful applications someone can suggest.

**Comment by nicm** [comment]
I'm not saying we can't add what you are asking for but I don't really see how a 5 second interval to check for the pane changing is better than a short delay when #() runs on your first pane.

You could run your main process all the time and have tmux query it with #(), it could use control mode to get notified when `#{pane_current_command}` changes. It could even use control mode to set a user option at that point which you could use from the status line, the  you wouldn't need #() at all.

**Comment by nicm** [comment]
Well the problem with control mode is you can only attach to one session, so you would need one for each session.

**Comment by mohkale** [comment]
> You could run your main process all the time and have tmux query it with #(), it could use control mode to get notified when #{pane_current_command} changes. It could even use control mode to set a user option at that point which you could use from the status line, the you wouldn't need #() at all.

I actually didn't know control-mode existed (the manual is so massive I must have missed it). I'll try and see whether I can do just this. Thank you for pointing it out, it looks very interesting :smile:.

> Well the problem with control mode is you can only attach to one session, so you would need one for each session.

I actually already do this, just so that if I have two tmux sessions active the icon for them matches pane you see in them, not the last active pane you interacted with. So I don't think this is an issue.

**Comment by nicm** [comment]
https://github.com/tmux/tmux/wiki/Control-Mode

You could look for `%sessions-changed` to spin up a new client and add a subscription to see when `pane_current_command` changes. You will need to run master or 3.2-rc however.

**Comment by mohkale** [comment]
@nicm 

Is there a way to attach to a client in command-mode without getting constant `%output` automatically getting printed? All the noise is making it hard to use it as a REPL for experimentation. `-CC` seems to disable input/output altogether :cry:.

**Comment by nicm** [comment]
Yes add `-fno-output`.

**Comment by mohkale** [comment]
Ooh, thanks for that. It should greatly reduce the amount of noise. I've managed to get a neet proof of concept working:

```sh
{ echo "refresh-client -B icon-tail::'#{pane_current_command}'"; cat } | tmux -C attach-session -t main -fno-output | grep '^%subscription-changed'
```

It reduces my manual delay to an instant command and can even suppress its output when the command hasn't changed. This is absolutely spectacular. I believe this issue can be closed now, thank you for your help :smile:.

**Comment by mohkale** [comment]
@nicm 

Is there a way to subscribe to whatever session a client is in rather than to the current session only. What I've got now works but when I switch my client to another session it seems to still only be tracking format changes from the previous session.

I have been using `#{client_tty}` as the argument to `-t`and I thought that means it would track the session the client was in but in reality I think it just finds the session the client is in and then doesn't update that session whenever the client switches sessions.

**Comment by nicm** [comment]
No the control mode client is its own client and has its own attached session. If you want it to track another client's session you will need to watch for `%client-session-changed` and change the control mode client's session to match with `switch-client`.

**Comment by mohkale** [comment]
Okay, thank you.

**Comment by mohkale** [comment]
@nicm 

Sorry to reopen this... but is there a way to detect when a client detaches/leaves in command-mode. At the moment I can track and follow a client well enough to show an icon for whatever pane/window their on... but when the client leaves I have no way to detect and kill my tracker and so I end up with a lot of zombie tracker processes.

**Comment by nicm** [comment]
I don't think so but it would be easy enough to add `%client-detached` or similar. Or you could poll `list-clients`.

**Comment by mohkale** [comment]
@nicm 

That's a shame. I'd rather add a %client-detached then poll list-clients. Could you walk me through the steps needed to implement?

**Comment by nicm** [comment]
There is already a client-detached hook so I think you should only need to connect that through from notify.c into control-notify.c in the same way as eg client-session-changed.

**Comment by nicm** [comment]
I can't look at the code right now but I can give you more details tomorrow if you're having trouble.

**Comment by mohkale** [comment]
@nicm 

Roger. I'll try to implement this now and get back to you if I have any issues. If it takes too long today I might put it off until next week.

**Comment by mohkale** [comment]
@nicm 

I've just noticed that if you kill the terminal that tmux is running in, you don't get a client detached message in control-mode. Is this intentional? I assume when tmux recieved HUP it exits itself but it doesn't seem to detach before exiting meaning the control-mode client isn't aware.

**Comment by nicm** [comment]
It isn't intentional but in that case it will go directly through `server_client_lost` rather than via `server_client_check_exit` first.

You could probably move the `notify_client` into the former. I'm not sure if this needs to be a separate notification type though to distinguish, not sure there is much use in that.

**Comment by mohkale** [comment]
@nicm 

I tried moving the client-detached message from `server_client_check_exit` to `server_client_lost` but there seems to be an extra client-detached message being outputted which I can't explain. If I run `:list-clients` beforehand the client tmux control mode shows as exited doesn't seem to exist.

**Comment by nicm** [comment]
I can take a look later in the week.

**Comment by mohkale** [comment]
@nicm 

Ping

**Comment by nicm** [comment]
Sorry, show me the changes you made that aren't working, I'll have a look.

**Comment by mohkale** [comment]
@nicm No worries, you can find the changes [here](https://github.com/mohkale/tmux/commit/ecd675d0ca8993f17e0995807637455aa52b0459)

**Comment by nicm** [comment]
Why did you remove the check for `CLIENT_ATTACHED`? I don't think we want to send messages for unattached clients.

With or without, I don't see any duplicate messages and it is hard to see how it could end up in `server_client_lost` twice for the same client. How do you reproduce?

**Comment by mohkale** [comment]
> Why did you remove the check for CLIENT_ATTACHED? I don't think we want to send messages for unattached clients.

oh, I assumed seeing as we were detaching a client that it was already attached (maybe my interpretation of the meaning there was flawed). Regardless I've added it back in and now I'm not seeing any extra messages :smile:.

**Comment by nicm** [comment]
No `server_client_lost` is called when a client exits for any reason, it may be before attach or it may be a client that never attaches (for example if you do `tmux lsw` from the shell it will be a client but it will not attach).

**Comment by nicm** [comment]
OK I have changed this now in OpenBSD, it'll be in GitHub later.
